### PR TITLE
(Proposal) Introducing a build step

### PR DESCRIPTION
**Pull request by gksander**
## Lightweight build-step

runpkg is dope, and I know it was originally built to be "build-step-less". In theory, I like the idea of not having a build step. However, I think most frontend devs are _used_ to having a build-step and the niceties it provides – such as TypeScript support, more explicit dependencies, etc.. I think that having tooling that more devs are used to using (such as standard React, TSX, etc.) will allow this project to attract more contributors, and make it easier for us to enhance and maintain going forward.

This PR is a draft at what that might look like. It includes:

- A setup for Vite, which requires _very little_ configuration and provides blazing fast builds. This supports TS out of the box, which is ❤️ 
- A Vite plugin for PWA stuff so that the service worker file can be transpiled (so that it can eventually be TS).
- `styled-components` for styling, which has an API close enough to `csz` that the conversion is easy enough. I'm proposing this because `styled-components` is more active and more familiar to many devs around this neck of the woods.
- A conversion of a handful of simple files from `.js` (and `htm`) to standard React with `.tsx`.

---

### Testing this out

You can test this branch out by pulling it down and running `yarn start`. Everything should still work as expected (though you might catch a few React warnings in the console that aren't present in the build-less step 👀 because we can run React in dev mode).

You can run `yarn build` to build the app to the `dist` folder, and you can test that out via `cd dist && npx http-server .`. Again, this built version should be running the same as the production site.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -13,30 +13,7 @@
...
-          htm: 'https://unpkg.com/htm@2.2.1/dist/htm.module.js',
-          csz: 'https://unpkg.com/csz@1.1.1/index.js',
-          'prism-react-renderer':
-            'https://unpkg.com/prism-react-renderer?module',
-          algolia:
-            'https://www.unpkg.com/algoliasearch@4.1.0/dist/algoliasearch-lite.esm.browser.js',
-        },
-      });
-      document.currentScript.after(im);
-    </script>
```
Comment (gksander) [inline index.html:38]: These are now npm-installed and imported directly where needed.

File: src/components/AlgoliaLogo.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import * as React from 'react';
+import styled from 'styled-components';
+
+export const AlgoliaLogo = () => {
```
Comment (gksander) [inline src/components/AlgoliaLogo.tsx:4]: Most of these new `.tsx` are basically just a renaming of the `.js` version, and modified slightly to use JSX.

File: src/components/Article.js [hunk 3]
```diff
@@ -17,9 +18,9 @@ export default () => {
         <article className=${styles.container}>
           <div className=${styles.welcome}>
             <a className=${styles.formidaLogo} href="https://formidable.com">
-              ${FormidableLogo}
+              <${FormidableLogo} />
```
Comment (gksander) [inline src/components/Article.js:21]: Components now converted to JSX can no longer be called in the shorthand manner.

File: src/components/Main.js [hunk 4]
```diff
@@ -1,12 +1,13 @@
...
 import { useStateValue } from '../utils/globalState.js';
 import { parseUrl } from '../utils/parseUrl.js';
+import AnalyzeFile from '../utils/analyzeFile.js?worker';
 
 import Nav from './Nav.js';
 import Article from './Article.js';
 
 const replaceState = url => history.replaceState(null, null, url);
-const worker = new Worker('./utils/analyzeFile.js');
+const worker = new AnalyzeFile();
```
Comment (gksander) [inline src/components/Main.js:10]: This is pretty dope. Vite supports webworkers out of the box, so we can eventually write this `analyzeFile` file in TS.

File: src/utils/analyzeFile.js [hunk 5]
```diff
@@ -1,16 +1,10 @@
...
 
 const fileNameRegEx = /\/[^\/@]+[\.][^\/]+$/;
 
-// Hack to get ESM into webworker as 21/Nov/2019 ESM is not supported by webworkers
-const getParseUrl = () =>
-  fetch('/utils/parseUrl.js')
-    .then(x => x.text())
-    .then(x =>
-      x.replace(/\s*export {[^}]+\};/g, '').replace(/^const\s[^=]+=/, '')
-    );
```
Comment (gksander) [inline src/utils/analyzeFile.js:12]: One of the benefits of build-step: don't have to hack this.

**Comment by robwalkerco** [comment]
@gksander why was this closed?
These changes would make it much easier to pull in [react-window](https://github.com/bvaughn/react-window) or similar to resolve #191

**Comment by gksander** [comment]
@robwalkerco My understanding is: this project had a goal outside of being just a package explorer: it was to experiment with building a "modern SPA" without needing a build step, to use modern browser ESM features to make this magic work.

To add a build step would be to erase some of the original motivation behind the project. But I have flirted with the idea of a similar project but with a build step.

**Pull request closed without merging**